### PR TITLE
Cleanup requires with macros

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This PR cleans the `requires*` meta programs to be auto generated using macros. This PR should be fully backwards compatible with the current method. More importantly, the documentation for these has been cleanup up such that it's much easier to go through and look for certain types of requires.

The macros are

`STAN_ADD_REQUIRE_UNARY`: For validating a single type satisfies a type trait.
`STAN_ADD_REQUIRE_BINARY`: For type traits that require comparing two types.
Both of these have a `*_INNER` macro that defines requires to check the `scalar_type` and `value_type` of arbitrary types. This is more useful for binary requires where we want to check that two matrices for instance have the same `value_type`. These are labeled as `require_vt/st_*` i.e. `require_vt_same<T, S>`

`STAN_ADD_REQUIRE_CONTAINER`: For defining requires like `require_eigen_t<>` and `require_eigen_vt<>`.

## Tests

All the current tests should pass. One important thing to check is the documentation via `make doxygen`. 

## Side Effects

These macros will be floating in the background, but I think they are named such that it would be hard to accidentally hit them.

## Release notes

Refactoring the `require_*` metaprograms to reduce code duplication and increase documentation.

## Checklist

- [x] Math issue #1739

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
@serban-nicusor-toptal it looks like the windows machines are taking a bit? Fine with it taking a while if there are a bunch of PRs but just a heads up

**Comment by serban-nicusor-toptal** [comment]
Hey, one machine seems down. Going to get it up now to speed up the queue. Thanks for the notice!

**Comment by SteveBronder** [comment]
@t4c1 would you have time to give this a lookover? It looks like a lot but it's mostly just moving stuff around and a good bit of docs. The only other change is that `value_type` and `scalar_type`'s eigen and std vector specializations were moved to `is_eigen` and `is_vector`. I think this makes sense because you can't detect an eigen or std vector's inner types until you can detect whether you have an eigen type or std vector type.

That and I updated the code for the functions which were giving warnings on doxygen so that we could use the warnings are errors flag in doxygen. Now if anyone does not doc correctly it will error out

File: stan/math/prim/meta/require_helpers.hpp [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+/*! @} */
+
+/** \ingroup macro_helpers
+ * Adds binary require aliases.
+ * @param check_type The name of the type to check, used to define
+ * `require_<check_type>_t`.
+ * @param checker A struct that returns holds a boolean `value`
+ * @param doxygen_group The doxygen group to add this requires to.
+ */
+#define STAN_ADD_REQUIRE_BINARY(check_type, checker, doxygen_group)          \
```
Comment (t4c1) [inline stan/math/prim/meta/require_helpers.hpp]: `STAN_ADD_REQUIRE_BINARY` and value/scalar versions seem to be unused. Is that intentional?

File: stan/math/prim/meta/require_helpers.hpp [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+/*! @} */
+
+/** \ingroup macro_helpers
+ * Adds unary require aliases that check the `scalar_type`.
+ * @param check_type The name of the type to check, used to define
+ * `require_<check_type>_t`.
+ * @param checker A struct that returns holds a boolean `value`
+ * @param doxygen_group The doxygen group to add this requires to.
+ */
+#define STAN_ADD_REQUIRE_UNARY_SCALAR(check_type, checker, doxygen_group)  \
```
Comment (t4c1) [inline stan/math/prim/meta/require_helpers.hpp]: If I understand the code correctly this is used to generate requires we did not have before this PR. This is fine, I just dont like how they are named. We now have `require_eigen_vt` and `require_var_vt`. In one case the check works on container and in other on its value. I think we should change the names of the new ones. Maybe `require_vt_var`?

Also `STAN_ADD_REQUIRE_UNARY_SCALAR` and `STAN_ADD_REQUIRE_UNARY_VALUE` are always used together. You can replace them with a single macro that does both.

File: test/unit/math/prim/meta/require_generics_test.cpp [hunk 3]
```diff
@@ -221,58 +221,6 @@ TEST(requires_prim_scal, var_or_arithmetic_test) {
...
-
-TEST(requires_prim_arr, generic_container_st_test) {
-  EXPECT_FALSE(
-      (require_container_st_tester<stan::is_vector, std::is_integral,
-                                   std::vector<std::vector<double>>>::value));
-  EXPECT_TRUE(
-      (require_container_st_tester<stan::is_vector, std::is_floating_point,
-                                   std::vector<std::vector<double>>>::value));
-}
-
```
Comment (t4c1) [inline test/unit/math/prim/meta/require_generics_test.cpp]: Why are those tests deleted?

File: stan/math/fwd/fun/sum.hpp [hunk 4]
```diff
@@ -14,19 +14,19 @@ namespace math {
...
  *
  * @tparam T type of elements in the vector
- * @param m Vector.
+ * @param v Vector.
  * @return Sum of vector entries.
  */
 template <typename T>
-inline fvar<T> sum(const std::vector<fvar<T>>& m) {
-  if (m.size() == 0) {
+inline fvar<T> sum(const std::vector<fvar<T>>& v) {
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/fwd/fun/sum.hpp]: These are just to fix the doxygen warnings but I  can undo them

File: doxygen/doxygen.cfg [hunk 5]
```diff
@@ -293,15 +293,6 @@ EXTENSION_MAPPING      =
...
 MARKDOWN_SUPPORT       = YES
 
-# When the TOC_INCLUDE_HEADINGS tag is set to a non-zero value, all headings up
-# to that level are automatically included in the table of contents, even if
-# they do not have an id attribute.
-# Note: This feature currently applies only to Markdown headings.
-# Minimum value: 0, maximum value: 99, default value: 0.
-# This tag requires that the tag MARKDOWN_SUPPORT is set to YES.
-
-TOC_INCLUDE_HEADINGS   = 0
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply doxygen/doxygen.cfg]: It throws a warning when compiling the docs that TOC_INCLUDE_HEADINGS is not a supported doxygen macro (I think it's for a later version)

File: README.md [hunk 6]
```diff
@@ -1,6 +1,8 @@
 The <b>Stan Math Library</b> is a C++, reverse-mode automatic differentiation library designed to be usable, extensive and extensible, efficient, scalable, stable, portable, and redistributable in order to facilitate the construction and utilization of algorithms that utilize derivatives.
 
-[![DOI](https://zenodo.org/badge/38388440.svg)](https://zenodo.org/badge/latestdoi/38388440)
+\htmlonly
+ <div><a href="https://zenodo.org/badge/latestdoi/38388440"><img src="https://zenodo.org/badge/38388440.svg"/></a></div>
+\endhtmlonly
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply README.md]: This makes the badge work on the site while getting rid of the doxygen compilation warning

File: stan/math/prim/meta.hpp [hunk 7]
```diff
@@ -45,5 +210,6 @@
 #include <stan/math/prim/meta/void_t.hpp>
 #include <stan/math/prim/meta/StdVectorBuilder.hpp>
 #include <stan/math/prim/meta/VectorBuilder.hpp>
+/** @}*/
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/prim/meta.hpp]: Yes I went through and add `@ingroup type_trait` to the type traits that were missing it

File: doxygen/doxygen.cfg [hunk 8]
```diff
@@ -881,7 +872,7 @@ EXAMPLE_RECURSIVE      = NO
 # that contain images that are to be included in the documentation (see the
 # \image command).
 
-IMAGE_PATH             =
+IMAGE_PATH             = doxygen/pretty_stuff
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply doxygen/doxygen.cfg]: Not needed will delete

File: stan/math/prim/meta/require_helpers.hpp (hunk 2 above)
Comment (SteveBronder) [reply stan/math/prim/meta/require_helpers.hpp]: > require_vt_var

I like it! In the past I've seen a few places where we accept any scalar/container type as long as it contains a var etc.  

> Also STAN_ADD_REQUIRE_UNARY_SCALAR and STAN_ADD_REQUIRE_UNARY_VALUE are always used together. You can replace them with a single macro that does both.

Will do

File: stan/math/prim/meta/require_helpers.hpp (hunk 1 above)
Comment (SteveBronder) [reply stan/math/prim/meta/require_helpers.hpp]: They are used in require_generics.hpp for comparison type traits such as 

```cpp
STAN_ADD_REQUIRE_BINARY(same, std::is_same, require_std);
```

File: stan/math/prim/meta.hpp [hunk 9]
```diff
@@ -1,6 +1,168 @@
...
+/**
+ * \ingroup require_meta
+ * \defgroup macro_helpers Require Macro Generators
+ * These macros are used on type traits to define the set of `requires`
+ */
+
+/** \addtogroup type_trait
+ *  @{
+ */
+
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/prim/meta.hpp]: Yeah I'll remove this

File: stan/math/fwd/fun/sum.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/fwd/fun/sum.hpp]: To undo these I will need to turn off the doxygen option to promote warnings to errors that I added in this PR

File: doxygen/doxygen.cfg [hunk 10]
```diff
@@ -2064,7 +2055,7 @@ PREDEFINED             = STAN_OPENCL \
 # definition found in the source code.
 # This tag requires that the tag ENABLE_PREPROCESSING is set to YES.
 
-EXPAND_AS_DEFINED      =
+EXPAND_AS_DEFINED      = STAN_ADD_REQUIRE_UNARY
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply doxygen/doxygen.cfg]: Thought I needed it but I dont' will delete

File: test/unit/math/prim/meta/require_generics_test.cpp (hunk 3 above)
Comment (SteveBronder) [reply test/unit/math/prim/meta/require_generics_test.cpp]: These were checking the `stan::require_container_vt` and `stan::require_container_st` which do not exist anymore

File: stan/math/fwd/fun/sum.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/fwd/fun/sum.hpp]: If that is so, I am ok with leaving those in. I did not understand these are connected with docs.

File: stan/math/prim/meta.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/prim/meta.hpp]: So this is now replaced with docs on each require(that are generated with macros)?

[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/prim/meta.hpp]: > is this just moved from stan/math.hpp?

I meant the whole ~150 lines of docs.

File: stan/math/prim/meta/require_generics.hpp [hunk 11]
```diff
@@ -1,976 +1,37 @@
...
+STAN_ADD_REQUIRE_UNARY(arithmetic, std::is_arithmetic,
+                       require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY_INNER(arithmetic, std::is_arithmetic,
+                             require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY(floating_point, std::is_floating_point,
+                       require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY_INNER(floating_point, std::is_floating_point,
+                             require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY(index, std::is_integral, require_stan_scalar_real);
+STAN_ADD_REQUIRE_UNARY_INNER(index, std::is_integral, require_stan_scalar_real);
```
Comment (t4c1) [inline stan/math/prim/meta/require_generics.hpp]: I know it is not part of this PR, but why do we have `require_index` for `is_integral`?

File: stan/math/fwd/fun/sum.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/fwd/fun/sum.hpp]: Actually, what kind of warnings did renaming a few variables solve?

[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/fwd/fun/sum.hpp]: I see you already reverted those. Lets leave them for another PR.

File: stan/math/prim/meta/require_generics.hpp (hunk 11 above)
Comment (SteveBronder) [reply stan/math/prim/meta/require_generics.hpp]: tbh I put that in there whenever I was working on stan upstream stuff, I can change this

**Comment by SteveBronder** [comment]
> Actually, what kind of warnings did renaming a few variables solve?

It's something weird in doxygen

```
stan/math/rev/fun/multiply.hpp:545: warning: argument 'm1' of command @param is not found in the argument list of stan::math::multiply(const Mat1 &A, const Mat2 &B)
```

**Comment by SteveBronder** [comment]
@t4c1 fyi I just went through and fixed index and some header includes rest of the stuff is just merge and a cpplint fix

https://github.com/stan-dev/math/pull/1794/commits/675b03d070b6019b64ce3f9f25ff794b95295e47

**Pull request merged**